mfe.go

```diff
@@ -0,0 +1,1092 @@
+}
+
+/**
+* EnergyContribution contains the energy contribution of a loop
+* The fields `loopType` and `energy` will always have a value for all loop types.
+* Except for `loopType` of `ExternalLoop`, all loops will have a
+* `closingFivePrimeIdx` and `closingThreePrimeIdx`.
+* Only interior loops will have `enclosedFivePrimeIdx` and `enclosedThreePrimeIdx`.
+ */
+type EnergyContribution struct {
+	closingFivePrimeIdx, closingThreePrimeIdx   int // The closing base pair is the base pair that closes the loop
+	loopType                                    int // The type of loop (ExternalLoop, InteriorLoop, HairpinLoop, or MultiLoop)
+	energy                                      int // The free energy contribution of the loop in dcal / mol
+	enclosedFivePrimeIdx, enclosedThreePrimeIdx int // The base pair (that is enclosed by a closing base pair) which delimits an interior loop (only available for interior loops)
+}
+
+const (
+	ExternalLoop = iota
+	InteriorLoop
+	HairpinLoop
+	MultiLoop
+)
+
+/**
+ *  CalculateMFE returns the free energy of an already folded RNA and a list of
+ *  energy contribution of each loop.
+ *
+ *  @param sequence         A RNA sequence
+ *  @param structure        Secondary structure in dot-bracket notation
+ *  @param temperature      Temperature at which to evaluate the free energy of the structure
+ *  @return                 The free energy of the input structure given the input sequence in kcal/mol
+ */
+func CalculateMFE(sequence, structure string, temperature float64) (float64, []EnergyContribution, error) {
+	lenSequence := len(sequence)
+	lenStructure := len(structure)
+
+	if lenSequence != lenStructure {
+		return 0, nil, fmt.Errorf("length of sequence (%v) != Length of structure (%v)", lenSequence, lenStructure)
+	} else if lenStructure == 0 {
+		return 0, nil, errors.New("lengths of sequence and structure cannot be 0")
+	}
+
+	sequence = strings.ToUpper(sequence)
+	// vivek: Should we convert DNA to RNA and should we ensure sequence is only
+	// made up of ACGU?
+
+	pairTable, err := pairTable(structure)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	fc := &foldCompound{
+		length:          lenSequence,
+		energyParams:    scaleEnergyParams(temperature),
+		sequence:        sequence,
+		encodedSequence: encodeSequence(sequence),
+		pairTable:       pairTable,
+	}
+
+	energyInt, energyContributions := evaluateFoldCompound(fc)
+	energy := float64(energyInt) / 100.0
+
+	return energy, energyContributions, nil
+}
+
+// Encodes a sequence into its numerical representaiton
+func encodeSequence(sequence string) []int {
+	// Make a nucleotide byte -> int map
+	nucleotideRuneMap := map[byte]int{
+		'A': 1,
+		'C': 2,
+		'G': 3,
+		'U': 4,
+	}
+
+	lenSequence := len(sequence)
+	encodedSequence := make([]int, lenSequence)
+
+	// encode the sequence based on nucleotideRuneMap
+	for i := 0; i < lenSequence; i++ {
+		encodedSequence[i] = nucleotideRuneMap[sequence[i]]
+	}
+
+	return encodedSequence
+}
```

Comment:
As we discussed, since LinearFold encodes nucleotides with a different map ('A' -> 0), we can't encode them with the same function. 

(Changing the MFE calculation code to use the same map as LinearFold would be a lot of work as this map dictates where the relevant energy values are found in the energy parameters matrices. So we'd have to change all those matrices if we change this mapping.)